chore(http): Improvements to HTTP messages page (#36029)
mdn/content: files/en-us/web/http/messages/index.md

---
title: HTTP messages
slug: Web/HTTP/Messages
page-type: guide
---

{{HTTPSidebar}}

> **HTTP messages** are how data is exchanged between a server and a client in the HTTP protocol.

Reviewer: Yuk.

```suggestion
**HTTP messages** are the mechanism used to exchange data between a server and a client in the HTTP protocol.
```

There are two types of messages: **requests** sent by the client to trigger an action on the server, and **responses**, the answer that the server sends in response to a request.

Developers rarely, if ever, build HTTP messages from scratch.
Applications such as a browser, a proxy, or a Web server, use software designed to create HTTP messages in a reliable and efficient way.
How messages are created or transformed is controlled via APIs in browsers, configuration files for proxies or servers, or other interfaces.

> In HTTP protocol versions up to HTTP/2, the message is text-based and it's relatively straightforward to read and understand after you've familiarized yourself with the format.

Reviewer: ```suggestion
In HTTP protocol versions up to HTTP/2, messages are text-based, and are relatively straightforward to read and understand after you've familiarized yourself with the format.
```

> In HTTP/2, messages are wrapped in binary framing, which makes them slightly harder to read without certain tools, but the underlying semantics of the protocol is the same.
> This means that you can learn the structure and meaning of HTTP messages based on the text-based format of HTTP/1.x messages, and apply this understanding to HTTP/2 and beyond.

Reviewer: ```suggestion
In HTTP/2, messages are wrapped in binary framing, which makes them slightly harder to read without certain tools.
However the underlying semantics of the protocol is the same, so you can learn the structure and meaning of HTTP messages based on the text-based format of HTTP/1.x messages, and apply this understanding to HTTP/2 and beyond.
```

> This guide uses HTTP/1.1 messages for readability, and explains the structure of HTTP messages in HTTP/1.1, highlighting some differences with HTTP/2 in the final section.

Reviewer: NOte sure if "with HTTP/2" would be better as "in"

Reviewer: What differences are we highlighting - in the text format, or in protocol. I think this might be what you mean (or something like):
```suggestion
This guide uses HTTP/1.1 messages for readability, and explains the structure of HTTP messages using the HTTP/1.1 format.
We highlight some differences that you might need for describing HTTP/2 in the final section.
```

Author: > What differences are we highlighting - in the text format, or in protocol.

In the protocol only, as HTTP/2 is binary framed as opposed to text-based

> [!NOTE]
> You can see HTTP messages in a browser's **Network** tab in the developer tools, or if you print HTTP messages to the console using CLI tools such as [curl](https://curl.se/), for example.

## Anatomy of an HTTP message

To understand how HTTP messages work, we'll look at HTTP/1.1 messages and examine the structure.

> The following illustration shows how messages in HTTP/1.1 look like and the similarities between requests and responses:

Reviewer: ```suggestion
The following illustration shows what messages in HTTP/1.1 look like and the similarities between requests and responses:
```

> ![Requests and responses share a common structure in HTTP](https://mdn.github.io/shared-assets/images/diagrams/http/messages/http-message-anatomy.svg)

Reviewer: Accept is a request header only - response should be Content-Type

Author: Well spotted, but appears we need an update to Accept:

https://httpwg.org/specs/rfc9110.html#field.accept

>When sent by a server in a response, Accept provides information about which content types are preferred in the content of a subsequent request to the same resource.

We can cover in this PR, actually

Both requests and responses share a similar structure:

> 1. A _start-line_ is a single line that describes the request or the outcome of a request.

Reviewer: Maybe. 

```suggestion
1. A _start-line_ is a single line that describes the HTTP version along with the request method or the outcome of the request.
```

Author: This is good, ty

> 2. An optional set of _HTTP headers_ with metadata that describes a message or how the client and server are communicating.

Reviewer: Don't like "how the client and server are communicating.". REmove altogether or be more precise. Maybe somethgn like.

```suggestion
2. An optional set of _HTTP headers_ containing metadata that describes the message. For example, a request for a resource might include the allowed formats of that resource, while the response might include headers to indicate the actual format returned.
```

3. An empty line indicating the metadata of the message is complete.

> 4. An optional _body_ containing data associated with the message or the resource associated with a response.

Reviewer: Again, perhaps no need to have text after "the message", or use a more concrete example of the kind of data. No
```suggestion
4. An optional _body_ containing data associated with the message. This might be POST data to send to the server in a request, or some resource returned to the client in a response.
```

   Whether a message contains a body or not is determined by the start-line and HTTP headers.

The start-line and headers of the HTTP message are collectively known as the _head_ of the requests, and the part afterwards that contains its content is known as the _body_.

## HTTP requests

Let's look at the following example HTTP POST request that's sent after a user submits a form on a web page:

```http
POST /users HTTP/1.1
Host: example.com
Content-Type: application/x-www-form-urlencoded
Content-Length: 50

name=FirstName%20LastName&email=bsmth%40example.com
```

The start-line in HTTP requests (`POST /users HTTP/1.1` in the example above) is called a "request-line" and is made of three parts:

```http
<method> <request-target> <protocol>
```

- `<method>`
  - : The [HTTP method](/en-US/docs/Web/HTTP/Methods) (also known as an _HTTP verb_) is one of a set of defined words that describes the meaning of the request and the desired outcome.
    For example, `GET` indicates that the client would like to receive a resource in return, and `POST` means that the client is sending data to a server.
- `<request-target>`

>   - : The request target is usually a {{glossary("URL")}} or a path and is characterized by the context of the request.

Reviewer: Maybe 
```suggestion
  - : The request target is usually an absolute or relative {{glossary("URL")}}, and is characterized by the context of the request.
```

>     The format of the request target depends on the HTTP method used and the request context, but is described in more detail in the [Request targets](#request-targets) section below

Reviewer: ```suggestion
    The format of the request target depends on the HTTP method used and the request context, and is described in more detail in the [Request targets](#request-targets) section below
```

- `<protocol>`

>   - : The _HTTP version_, which defines the structure of the remaining message, acting as an indicator of the expected version to use for the response.

Reviewer: List the main options?

### Request targets

There are a few ways of describing a request target, but by far the most common is the "origin form".
Here's a list of the types of targets and when they are used:

1. In _origin form_, the recipient combines an absolute path with the information in the {{HTTPHeader("Host")}} header.

>    A query string can be appended to the path for additional information (usually in `key=value` format).

Reviewer: Usually or always?

Author: I think usually, since you can have keys only, like:

```
?key1&key2
```

And may be niche cases to use some other kind of custom format, I'm sure.

https://datatracker.ietf.org/doc/html/rfc3986#section-3.4

   This is used with `GET`, `POST`, `HEAD`, and `OPTIONS` methods:

   ```http
   GET /en-US/docs/Web/HTTP/Messages HTTP/1.1
   ```

2. The _absolute form_ is a complete URL, including the authority, and is used with `GET` when connecting to a proxy:

   ```http
   GET https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages HTTP/1.1
   ```

3. The _authority form_ is the authority and port separated by a colon (`:`).
   It is only used with the {{HTTPMethod("CONNECT")}} method when setting up an HTTP tunnel:

   ```http
   CONNECT developer.mozilla.org:443 HTTP/1.1
   ```

4. The _asterisk form_ is only used with `OPTIONS` when you want to represent the server as a whole (`*`) as opposed to a named resource:

   ```http
   OPTIONS * HTTP/1.1
   ```

> ### Request headers

Author: I think the original is okay, it's probably likely we'll have people searching for "request headers", what do you reckon?

> ### Request headers

Reviewer: Definitely OK.

Headers are metadata sent with a request after the start line and before the body.
In the [form submission example](#http_requests) above, they are the following lines of the message:

```http
Host: example.com
Content-Type: application/x-www-form-urlencoded
Content-Length: 50
```

> In HTTP/1.x, they are a **case-insensitive** string followed by a colon (`:`) and a value whose format depends on the header.

Reviewer: ```suggestion
In HTTP/1.x, each header is a **case-insensitive** string followed by a colon (`:`) and a value whose format depends on the header.
```

> The whole header, including the value, consists of one single line, which can be quite long for some cases like the {{HTTPHEader("Cookie")}} header.

Reviewer: ```suggestion
The whole header, including the value, consists of one single line.
This line can be quite long in some cases, such as the {{HTTPHEader("Cookie")}} header.
```

![Example of headers in an HTTP request](https://mdn.github.io/shared-assets/images/diagrams/http/messages/request-headers.svg)

> There are many different headers that can appear in requests, but they are categorized as follows:

Reviewer: As per comment above, https://github.com/mdn/content/pull/36029/files#r1796474253 it might be best to be clear about our headers in the request

```suggestion

Some headers in a request message are explicitly referred to as request headers. Others can be sent in both requests and responses, or might have a more fine-grained categorization:
```

Reviewer: I only got this far in review @bsmth . Will try again Monday.

Author: I see what you mean. What do you think about this:

>Some headers are exclusively used in requests, others can be sent in both requests and responses, or might have a more specific categorization:

Reviewer: Sure :-)

Reviewer: ```suggestion
Some headers are exclusively used in requests, while others can be sent in both requests and responses, or might have a more specific categorization:
```

In the [form submission example](#http_requests), this part is the body:

```http
name=FirstName%20LastName&email=bsmth%40example.com
```

The body in the form submission request contains a relatively small amount of information as `key=value` pairs, but a request body could contain other types of data that the server expects:

```json
{
  "firstName": "Brian",
  "lastName": "Smith",
  "email": "bsmth@example.com",
  "more": "data"
}
```

or data in multiple parts:

```http
--delimiter123
Content-Disposition: form-data; name="field1"

value1
--delimiter123
Content-Disposition: form-data; name="field2"; filename="example.txt"

Text file contents
--delimiter123--
```

## HTTP responses

Responses are the HTTP messages a server sends back in reply to a request.
The response lets the client know what the outcome of the request was.
Here's an example HTTP/1.1 response to a `POST` request that created a new user:

```http
HTTP/1.1 201 Created
Content-Type: application/json
Location: http://example.com/users/123

{
  "message": "New user created",
  "user": {
    "id": 123,
    "firstName": "Example",
    "lastName": "Person",
    "email": "bsmth@example.com"
  }
}
```

> The start-line (`HTTP/1.1 201 Created` above) is called a "status line" in responses and is made of three parts:

Reviewer: ```suggestion
The start-line (`HTTP/1.1 201 Created` above) is called a "status line" in responses, and has three parts:
```

```http
<protocol> <status-code> <status-text>
```

- `<protocol>`
  - : The _HTTP version_ of the remaining message.
- `<status-code>`
  - : A numeric [status code](/en-US/docs/Web/HTTP/Status) that indicates whether the request succeeded or failed.
    Common status codes are {{HTTPStatus("200")}}, {{HTTPStatus("404")}}, or {{HTTPStatus("302")}}.
- `<status-text>`
  - : The status text is a brief, purely informational, textual description of the status code to help a human understand the HTTP message.

### Response headers

Response headers are the metadata sent with a response.

> In HTTP/1.x, they are a **case-insensitive** string followed by a colon (`:`) and a value whose format depends upon which header is used.

Reviewer: ```suggestion
In HTTP/1.x, each header is a **case-insensitive** string followed by a colon (`:`) and a value whose format depends upon which header is used.
```

> ![Example of headers in an HTTP response](https://mdn.github.io/shared-assets/images/diagrams/http/messages/response-headers.svg)

Reviewer: Oh, didn't know ETag was a representation header - It isn't marked as such in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag

I guess it should be if it describes the state of the data at a particular time.

Author: Yeah this is really tricky, and I've found bugs online where developers have stumbled on this, the spec says that you should get different Etags for different representations (compression applied by a load balancer, etc.)

https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#strong_validation

We can add this to the ETag page

Author: added in 60cd7e726f44f8f52295bfe6b679c970cf75fb63

Like request headers, there are many different headers that can appear in responses, and they are categorized as:

> - {{glossary("Response header", "Response headers")}} give additional information about the server which doesn't fit in the status line.

Reviewer: True, but perhaps you might say a little more - i.e. it is also about the message - such as how the message should be cached.

Author: what do you think about this one:

60cd7e726f44f8f52295bfe6b679c970cf75fb63

Reviewer: Yes, certainly gives more of a flavour.

- {{glossary("Representation header", "Representation headers")}} sent in a response if the message has a body.
  They describe the original form of the message data and any encoding applied.
  This allows a recipient to understand how to reconstruct the resource as it was before it was transmitted over the network.

### Response body

A response body is included in most messages when responding to a client.
In successful requests, the response body contains the data that the client asked for in a `GET` request.

> If there are problems with the client's request, it's common that the response body describes why the request failed, with some hints about whether it's permanent or temporary, for example.

Reviewer: I don't think we need "for example"

```suggestion
If there are problems with the client's request, it's common for the response body to describe why the request failed, and hint as to whether it's permanent or temporary.
```

Response bodies may be:

- Single-resource bodies defined by the two headers: {{HTTPHeader("Content-Type")}} and {{HTTPHeader("Content-Length")}}, or of unknown length and encoded in chunks with {{HTTPHeader("Transfer-Encoding")}} set to `chunked`.
- [Multiple-resource bodies](/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#multipartform-data), consisting of a body that contains multiple parts, each containing a different piece of information.
  Multipart bodies are typically associated with [HTML Forms](/en-US/docs/Learn/Forms), but may also be sent in response to [Range requests](/en-US/docs/Web/HTTP/Range_requests).

> Responses with a status code that answers the request without the need to include message content like {{HTTPStatus("201", "201 Created")}} or {{HTTPStatus("204", "204 No Content")}} do not have a body.

Reviewer: ```suggestion
Responses with a status code that answers the request without the need to include message content, such as {{HTTPStatus("201", "201 Created")}} or {{HTTPStatus("204", "204 No Content")}}, do not have a body.
```

## HTTP/2 messages

> HTTP/1.x uses text-based messages that are straightforward to read and construct, but there are a few downsides to this.

Reviewer: ```suggestion
HTTP/1.x uses text-based messages that are straightforward to read and construct, but as a result have a few downsides.
```

You can compress message bodies using `gzip` or other compression algorithms, but not headers.
Headers are often similar or identical in a client-server interaction, but they are repeated in successive messages on a connection.
There are many known methods to compress repetitive text that are very efficient, which leaves a large amount of bandwidth savings unutilized.

> HTTP/1.x also has a problem called Head-of-Line blocking (HOL blocking) on the TCP connection, where a client has to wait for a response from the server before sending the next request.

Reviewer: ```suggestion
HTTP/1.x also has a problem called Head-of-Line (HOL) blocking on the TCP connection, where a client has to wait for a response from the server before sending the next request.
```

HTTP [pipelining](/en-US/docs/Web/HTTP/Connection_management_in_HTTP_1.x#http_pipelining) tried to work around this, but poor support and complexity means it's rarely used and difficult to get right.
Several connections need to be opened to send requests concurrently; and warm (established and busy) connections are more efficient than cold ones due to TCP slow start.

In HTTP/1.1 if you want to make two requests in parallel, you have to open two connections:

![Making two HTTP requests to a server in parallel](https://mdn.github.io/shared-assets/images/diagrams/http/messages/http-1-connection.png)

> This means that browsers are limited in the number of resources that they can download and render at the same time, which has typically been limited to 6 parallel connections in browsers.

Reviewer: ```suggestion
This means that browsers are limited in the number of resources that they can download and render at the same time, which has typically been limited to 6 parallel connections.
```

In the lines after the header transmission, you will see the following lines:

```http
[  0.447] send HEADERS frame ...
          ...
          :method: GET
          :path: /
          :scheme: https
          :authority: www.example.com
          accept: */*
          accept-encoding: gzip, deflate
          user-agent: nghttp2/1.61.0
```

This should look familiar if you're already comfortable working with HTTP/1.x and the concepts covered in the earlier section of this guide still apply.
This is the binary frame with the `GET` request for `example.com`, converted into a readable form by `nghttp`.
If you look further down the output of the command, you will see the `:status` pseudo-header in one of the streams received from the server:

```http
[  0.433] recv (stream_id=13) :status: 200
[  0.433] recv (stream_id=13) content-encoding: gzip
[  0.433] recv (stream_id=13) age: 112721
[  0.433] recv (stream_id=13) cache-control: max-age=604800
[  0.433] recv (stream_id=13) content-type: text/html; charset=UTF-8
[  0.433] recv (stream_id=13) date: Fri, 13 Sep 2024 12:56:07 GMT
[  0.433] recv (stream_id=13) etag: "3147526947+gzip"
...
```

> And if you remove the timing and stream ID from this message, it should start looking familiar:

Reviewer: ```suggestion
And if you remove the timing and stream ID from this message, it should even more familiar:
```

```http
:status: 200
content-encoding: gzip
age: 112721
```

Digging further into message frames, stream IDs and how the connection is managed is beyond the scope of this guide, but for the purpose of understanding and debugging HTTP/2 messages, you should be well-equipped using the knowledge and tools in this article.

## Conclusion

This guide gives a general overview of the anatomy of HTTP messages from HTTP/1.X to HTTP/2.
If you've followed the examples and understood the concepts described, you should be able to work with HTTP and understand how applications can use the protocol to send and receive data.

> The HTTP/2 framing mechanism adds a new intermediate layer between the HTTP/1.x syntax and the underlying transport protocol, without fundamentally modifying it, which builds upon proven mechanisms.

Reviewer: This line is dangling. If you want to keep it, swap the last two lines in the paragraph.

Are we going to talk about HTTP/3 ?

Author: >Are we going to talk about HTTP/3 ?

We should, I think we can keep it general for now, I've had a crack at some of the major points (why do we need major revisions) without going too deep into the weeds:

6a6ee7b3c0e4bf4a5be599d3c4e170a3af258ea1

Reviewer: Looks good to me

[  0.433] recv (stream_id=13) age: 112721
[  0.433] recv (stream_id=13) cache-control: max-age=604800
[  0.433] recv (stream_id=13) content-type: text/html; charset=UTF-8
[  0.433] recv (stream_id=13) date: Fri, 13 Sep 2024 12:56:07 GMT
[  0.433] recv (stream_id=13) etag: "3147526947+gzip"
...
```

And if you remove the timing and stream ID from this message, it should be even more familiar:

```http
:status: 200
content-encoding: gzip
age: 112721
```

Digging further into message frames, stream IDs and how the connection is managed is beyond the scope of this guide, but for the purpose of understanding and debugging HTTP/2 messages, you should be well-equipped using the knowledge and tools in this article.

## Conclusion

This guide provides a general overview of the anatomy of HTTP messages, using the HTTP/1.1 format for illustration.
We also explored HTTP/2 message framing, which introduces a layer between the HTTP/1.x syntax and the underlying transport protocol without fundamentally modifying HTTP's semantics.
HTTP/2 was introduced to solve the head-of-line blocking issues present in HTTP/1.x by enabling multiplexing of requests.

> One issue that remained in HTTP/2 is that head-of-line blocking still exists within TCP, so a performance bottleneck is now at the transport, rather than protocol, level.

Reviewer: Just felt a bit clunky

```suggestion
One issue that remained in HTTP/2 is that even though head-of-line blocking was fixed in the protocol level, there is still a performance bottleneck due to head-of-line blocking within TCP (at the transport level).
```
